wire-ios-sync-engine/Source/Data Model/Conversation+AccessMode.swift

```diff
@@ -228,30 +217,32 @@ extension ZMConversation {
             return completion(.failure(SetAllowServicesError.unknown))
         }
 
-        setAllowGuestsAndServices(allowGuests: self.allowGuests, allowServices: allowServices, in: userSession, apiVersion: apiVersion, completion)
+        setAllowGuestsAndServices(
+            allowGuests: self.allowGuests,
+            allowServices: allowServices,
+            apiVersion: apiVersion,
+            completion
+        )
 
     }
 
     /// Changes the conversation access mode to allow services.
-    private func setAllowGuestsAndServices(allowGuests: Bool, allowServices: Bool, in userSession: ZMUserSession, apiVersion: APIVersion, _ completion: @escaping (Result<Void, Error>) -> Void) {
-        let request = WirelessRequestFactory.setAccessRoles(allowGuests: allowGuests, allowServices: allowServices, for: self, apiVersion: apiVersion)
+    private func setAllowGuestsAndServices(allowGuests: Bool, allowServices: Bool, apiVersion: APIVersion, _ completion: @escaping (Result<Void, Error>) -> Void) {
 
-        request.add(ZMCompletionHandler(on: managedObjectContext!) { response in
-            if let payload = response.payload,
-               let event = ZMUpdateEvent(fromEventStreamPayload: payload, uuid: nil) {
-                self.allowGuests = allowGuests
-                self.allowServices = allowServices
+        var action = SetAllowGuestsAndServicesAction(
+            allowGuests: allowGuests,
+            allowServices: allowServices,
+            conversationID: self.objectID
+        )
 
-                // Process `conversation.access-update` event
-                userSession.processUpdateEvents([event])
+        action.perform(in: self.managedObjectContext!.notificationContext) { result in
```

Comment:
suggestion: safely unwrap the context and return an error if needed.

Reply:
Done!